src/SA/ldapsearch/entry.c

```diff
@@ -0,0 +1,398 @@
+DWORD GetTimeInSeconds(VOID)
+{
+    LARGE_INTEGER Time;
+    FILETIME FileTime;
+    DWORD dwSeconds;
+    int junk = 0;
+    lldiv_t test;
+    KERNEL32$GetSystemTimeAsFileTime(&FileTime);
+    Time.u.LowPart = FileTime.dwLowDateTime;
+    Time.u.HighPart = FileTime.dwHighDateTime;
+    Time.QuadPart = Time.QuadPart - TICKSTO1970;
+    Time.QuadPart = (uint64_t)Time.QuadPart / (uint64_t)TICKSPERSEC;
+    dwSeconds = Time.u.LowPart;
+
+    return dwSeconds;
+}
+#endif
+VOID ReturnDateTime(DWORD dwSeconds)
+{
+    LARGE_INTEGER Time;
+    FILETIME FileTime;
+    SYSTEMTIME SystemTime;
+    PCHAR DateBuffer[80];
+    PCHAR TimeBuffer[80];
+
+    Time.QuadPart = ((LONGLONG)dwSeconds * TICKSPERSEC) + TICKSTO1970;
+    
+    FileTime.dwLowDateTime = Time.u.LowPart;
+    FileTime.dwHighDateTime = Time.u.HighPart;
+    KERNEL32$FileTimeToLocalFileTime(&FileTime, &FileTime);
+    KERNEL32$FileTimeToSystemTime(&FileTime, &SystemTime);
+
+    KERNEL32$GetDateFormatW(LOCALE_USER_DEFAULT,
+                   DATE_SHORTDATE,
+                   &SystemTime,
+                   NULL,
+                   DateBuffer,
+                   ARRAYSIZE(DateBuffer));
+
+    KERNEL32$GetDateFormatW(LOCALE_USER_DEFAULT,
+                   TIME_NOSECONDS,
+                   &SystemTime,
+                   NULL,
+                   TimeBuffer,
+                   ARRAYSIZE(TimeBuffer));
+
+    internal_printf("%S %S", DateBuffer, TimeBuffer);
+}
+
+PCHAR GetDomainControllerHostName(){
+	DWORD dwRet;
+	PDOMAIN_CONTROLLER_INFO pdcInfo;
+
+	dwRet = NETAPI32$DsGetDcNameA(NULL, NULL, NULL, NULL, 0, &pdcInfo);
+	if (ERROR_SUCCESS == dwRet) {
+		internal_printf("[*] Using DC: %s\n", pdcInfo->DomainControllerName);		
+	} else {
+ 	    BeaconPrintf(CALLBACK_ERROR, "Failed to identify PDC, are we domain joined?");
+	}
+
+	PCHAR hostName = pdcInfo->DomainControllerName + 2; // Remove preceeding backslashes
+
+	NETAPI32$NetApiBufferFree(pdcInfo);
+	return hostName;
+}
+
+LDAP* InitialiseLDAPConnection(PCHAR hostName, PCHAR distinguishedName){
+	LDAP* pLdapConnection = NULL;
+
+    pLdapConnection = WLDAP32$ldap_init(hostName, 389);
+    
+    if (pLdapConnection == NULL)
+    {
+      	BeaconPrintf(CALLBACK_ERROR, "Failed to establish LDAP connection on 389.");
+        WLDAP32$ldap_unbind(pLdapConnection);
+    }
+    else {
+		internal_printf("[+] Initialised Connection.\n");
+    }
+
+	//////////////////////////////
+	// Bind to DC
+	//////////////////////////////
+    ULONG lRtn = 0;
+
+    lRtn = WLDAP32$ldap_bind_s(
+                pLdapConnection,      // Session Handle
+                distinguishedName,                // Domain DN
+                NULL,                 // Credential structure
+                LDAP_AUTH_NEGOTIATE); // Auth mode
+
+    if(lRtn == LDAP_SUCCESS)
+    {
+    	internal_printf("[+] Bind Successful.\n");
+    }
+    else
+    {
+      	BeaconPrintf(CALLBACK_ERROR, "Bind Failed: %i", lRtn);
```

Comment:
lRtn is a ULONG, the formatter should be %u or %x to match it being unsigned otherwise a negative number may be printed which will be difficult for the user to look up